### PR TITLE
[LEVWEB-1393] Maintain shape when redacting

### PR DESCRIPTION
**Pull request by aarongill1**
Function added to redact data that works with blocked records. As discussed, redacted and blocked fields now are given a value of 'null', rather than different 'shapes' for un-redacted, redacted and blocked records.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lev-api",
-  "version": "0.13.0",
+  "version": "0.14.0",
```
Comment (daniel-ac-martin) [inline package.json:3]: I think this can stay as `v0.13`. i.e. `0.13.1`.

Comment (aarongill1) [reply package.json]: Done

File: src/lib/censorDeath.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    coroner: {
+      name: censorField(true, death.coroner.name),
+      designation: censorField(true, death.coroner.designation),
+      area: censorField(true, death.coroner.area)
+    },
+    inquestDate: censorField(true, death.inquestDate),
+    status: {
+      blocked: death.status.blocked,
+      correction: censorField(true, death.status.correction),
+      marginalNote: censorField(true, death.status.marginalNote),
```
Comment (daniel-ac-martin) [inline src/lib/censorDeath.js:67]: I don't think these two should be redacted. They are required for the flags.

File: src/lib/censorDeath.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      onAuthorityOfRegistrarGeneral: censorField(true, death.status.onAuthorityOfRegistrarGeneral)
+    },
+    previousRegistration: death.previousRegistration,
+    nextRegistration: death.nextRegistration
```
Comment (daniel-ac-martin) [inline src/lib/censorDeath.js:71]: I think we need to run `censorDeath` on these if they exist.
e.g.
```js
    previousRegistration: death.previousRegistration && censorDeath(death.previousRegistration, roles),
```

File: test/unit/lib/censorDeath.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    occupation: "Mother occupation"
+  },
+  status: {
+    blocked: false,
+    correction: "None",
+    marginalNote: "None",
+    onAuthorityOfRegistrarGeneral: false
+  },
+  previousRegistration: {},
+  nextRegistration: {}
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:83]: I don't think these are valid values.
Better to set them to `undefined` I think.

File: test/unit/lib/censorDeath.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    occupation: null
+  },
+  status: {
+    blocked: true,
+    correction: null,
+    marginalNote: null,
+    onAuthorityOfRegistrarGeneral: null
+  },
+  previousRegistration: {},
+  nextRegistration: {}
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:159]: Same here.

File: test/unit/lib/censorDeath.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    occupation: null
+  },
+  status: {
+    blocked: false,
+    correction: null,
+    marginalNote: null,
+    onAuthorityOfRegistrarGeneral: null
+  },
+  previousRegistration: {},
+  nextRegistration: {}
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:239]: Same here.

File: test/unit/lib/censorDeath.js [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+describe('lib/censorDeath.js', () => {
+  describe('censorDeath function', () => {
+    it('should be a function', () => {
+      expect(censorDeath).to.be.a('function')
+    })
+    describe('record not blocked', () => {
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:247]: Better to write some nicer English here. Perhaps `'when the record is NOT blocked'`.

File: test/unit/lib/censorDeath.js [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+      it('should return the full death data if role has full access', () => {
+        expect(censorDeath(Object.assign({}, deathData), roleWithFullAccess))
+          .to.deep.equal(deathData);
+      })
+      it('should return redacted death data if role has not got full access', () => {
+        expect(censorDeath(Object.assign({}, deathData), roleWithoutFullAccess))
+          .to.deep.equal(redactedDataNotBlocked);
+      })
+    })
+    describe('record blocked', () => {
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:257]: Similar issue here.

File: test/unit/lib/censorDeath.js (hunk 7 above)
Comment (daniel-ac-martin) [reply test/unit/lib/censorDeath.js:247]: The reason being that we want some poor future programmer to be able to understand what has gone wrong as soon as possible.

**Review by daniel-ac-martin — CHANGES_REQUESTED** [review]
Looks like it needs a couple of changes but it's looking pretty good otherwise.

File: src/lib/censorDeath.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    },
+    inquestDate: censorField(true, death.inquestDate),
+    status: {
+      blocked: death.status.blocked,
+      correction: censorField(false, death.status.correction),
+      marginalNote: censorField(false, death.status.marginalNote),
+      onAuthorityOfRegistrarGeneral: censorField(true, death.status.onAuthorityOfRegistrarGeneral)
+    },
+    ...!death.nextRegistration && { previousRegistration: death.previousRegistration && censorDeath(death.previousRegistration, roles) },
+    ...!death.previousRegistration && { nextRegistration: death.nextRegistration && censorDeath(death.nextRegistration, roles) },
```
Comment (daniel-ac-martin) [inline src/lib/censorDeath.js:71]: Unfortunately, we can't detect whether we are in the 'next' chain or the 'previous' chain like this.

The base record be in the middle of the chain and this code would remove the links to the records.

I think you will need to do something like:
```js
const r = {[...]}

if (death.nextRegistration) {
  r.nextRegistration = censorDeath(death.nextRegistration, roles)
}

if (death.previousRegistration) {
  r.previousRegistration = censorDeath(death.previousRegistration, roles)
}

return r;

Comment (daniel-ac-martin) [reply src/lib/censorDeath.js:71]: We may even need to add some logic to distinguish between `null` and `undefined`. (Although probably not.)

File: test/unit/lib/censorDeath.js [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    occupation: "Mother occupation"
+  },
+  status: {
+    blocked: false,
+    correction: "None",
+    marginalNote: "None",
+    onAuthorityOfRegistrarGeneral: false
+  },
+  previousRegistration: undefined,
+  nextRegistration: undefined
```
Comment (daniel-ac-martin) [inline test/unit/lib/censorDeath.js:83]: Now that I think about it, these should probably be `null` or simply deleted.

**Review by daniel-ac-martin — CHANGES_REQUESTED** [review]
Very close, but some of the logic around the registration 'chain' needs updating.

File: test/unit/lib/censorDeath.js (hunk 10 above)
Comment (aarongill1) [reply test/unit/lib/censorDeath.js:83]: Changed to null

File: src/lib/censorDeath.js (hunk 9 above)
Comment (aarongill1) [reply src/lib/censorDeath.js]: Updated

File: src/lib/censorDeath.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
+const censorDeath = (death, roles) => {
+  const r = {};
+  if (death.nextRegistration) {
+    r.nextRegistration = censorDeath(death.nextRegistration, roles)
+  }
+  else if (death.previousRegistration) {
```
Comment (daniel-ac-martin) [inline src/lib/censorDeath.js:6]: `else if` creates the same problem as before.

Sometimes you have both `nextRegistration` and `previousRegistration`, you need to be able to deal with that.

**Comment by daniel-ac-martin** [comment]
Closing in favour of #74.

**Pull request closed without merging**